### PR TITLE
MAINT: interpolate: deduplicate input validation for NdBSpline

### PR DESCRIPTION
**Pull request by ev-br**
1. always store tuples of knots as 2D arrays internally
2. centralize input validations in `NdBSpline.__call__` and `.design_matrix`
3. pre-compute what we can in `__init__` instead of `__call__`
4. use properties for knots and degrees: user-facing API is tuples, so hide internal details.

Split off #21551 to keep python-level and compiled code refactorings separate. As a side effect, this tightens the `design_matrix` input validation a bit. Unlikely to cause any user impact: this is all very new and most user-facing usage is through RGI anyhow.

EDIT: This has barely any new code: despite what GH shows, this mostly moves code around.

### CODE REVIEW

File: scipy/interpolate/_ndbspline.py
Comment (ev-br) [inline scipy/interpolate/_ndbspline.py:223]:
```suggestion
                f" {xvals.shape[-1] = }."
```

File: scipy/interpolate/_ndbspline.py
Comment (j-bowhay) [inline scipy/interpolate/_ndbspline.py:271]: possibly `hasattr(k,  "__len__")` instead?

File: scipy/interpolate/_ndbspline.py
Comment (j-bowhay) [inline scipy/interpolate/_ndbspline.py:318]: Again feel free to ignore as it's just a copy paste from elsewhere but
```suggestion
    _t = np.full((ndim, max(len_t)), np.nan, dtype=float)
```

**Review by j-bowhay — APPROVED** [review]
LGTM, a couple of minor nits if you're touching this code again soon but won't block this PR on them since it's just a copy paste

**Pull request merged**
